text/3723-static-function-argument-unpacking.md

```diff
@@ -0,0 +1,719 @@
+```rust
+// At call-site, there seem to be three comma-separated places,
+// while the function has five parameter slots
+takes_five(0, ...tup3, 4)
+    
+// Desugared, the call above looks like this:
+takes_five(0, tup3.0, tup3.1, tup3.2, 4);
+//            ^^^^^^^^^^^^^^^^^^^^^^^
+// The whole tuple of three fields was unpacked into the
+// underlined parameter slots.
+}
+```
+
+## Non-Trivial Cases
+
+If, inside the call parentheses, the collection's fields can currently be accessed manually, in order, with `.idx`/`[idx]`, entering each as arguments to consecutive parameter slots, unpacking is valid.
+
+### Generic Parameters
+
+When function parameters are generic, using `<T>`, `impl` or `dyn`, exactly the same should happen as when the arguments are passed by hand. I.e., the argument's type must be *compatible* with the parameter's type. Just as when entering that argument manually with a field access expression.
+
+### References and Mutability in Function Parameters
+
+Rust's current syntax for calling functions that take one parameter by value, by reference, or by mutable reference requires the developer to be explicit about what they want:
+```rust
+fn ret_one_arg() -> u8 {
+    123
+}
+
+fn use_one_arg(p: u8) {
+    println!("{p}");
+}
+
+fn use_one_refarg(p: &u8) {
+    println!("{p}");
+}
+
+fn use_one_refmutarg(p: &mut u8) {
+    println!("{p}");
+}
+
+fn main() {
+    use_one_arg(ret_one_arg());
+    use_one_refarg(&ret_one_arg()); // `&` needed to compile!
+    use_one_refmutarg(&mut ret_one_arg()); // `&mut` needed to compile!
+}
+```
+
+Rust also requires the developer to explicitly dereference references:
+```rust
+const CONST_NUMBER: u8 = 42;
+
+fn ret_one_refarg() -> &'static u8 {
+    &CONST_NUMBER
+}
+
+fn use_one_arg(p: u8) {
+    println!("{p}");
+}
+
+fn use_one_refarg(p: &u8) {
+    println!("{p}");
+}
+
+fn main() {
+    use_one_arg(*ret_one_refarg()); // `*` needed to compile!
+    use_one_refarg(ret_one_refarg());
+}
+```
+
+Explicitly indicating varying degrees of (de)reference status or mutability on arguments being unpacked does not follow from the proposed syntax in any straightforward way. Thus, although it limits the usefulness of the feature, the design for such possibility is left out of the scope of this proposal. Consequently, the code will only compile if passing the arguments one by one with the corresponding field access expressions would compile.
+
+### Type Coercions of Collections
+
+If the collection being unpacked is a reference for the collection type, whether argument unpacking works, depends on if accessing it directly with the field access expression (`.idx`, or `[idx]`) would work at compile time. If it does, then argument unpacking works. (For the reference, see [`std::ops::Deref`](https://doc.rust-lang.org/std/ops/trait.Deref.html) and [type coercions](https://doc.rust-lang.org/reference/type-coercions.html).)
```

Comment:
Earlier it is stated that "Tuples, tuple structs, and fixed-size arrays can be unpacked.", but this seems to indicate that anything implementing `Index<Int>` could be unpacked. I think tuples definitely make sense and fixed-length `[T; N]` arrays could make sense, but anything involving `Index` should be out of scope for now because it turns the complexity up quite a bit.

It is pretty easy to turn slicelike collections into fixed-length arrays anyway, the user can handle that if they need this support.

Reply:
On a similar note, it is probably worth mentioning type inference. For example:

```rust
fn foo(a: u8, b: u8, c: u8, d: u8, e: u8) { /* ... */ }

fn bar(buf: &[u8]) {
    foo(...buf.try_into().unwrap());
}
```

Is that expected to work because `&[T; 5]` has `TryFrom<&[T]>` and `u8` is `Copy`? What about

```rust
fn bar(buf: &[u8], other: &[u8]) {
    foo(1, ...buf.try_into().unwrap(), ...other.try_into().unwrap());
}
```

That probably needs to be forbidden as ambiguous, but that is something that should be spelled out here with some T-Types input.

Reply:
Good catch – I seem to have written this part about accessing the fields in an ambiguous and self-contradictory way. I'll need to make some changes and also study the problem a bit more.

On the whole, I'm aiming at a design that's infallible during compilation. But I think I'll need to look into if that claim can actually be made after all, i.e. if some seemingly infallible cases wouldn't be totally infallible, but only *just as infallible* as they currently are. (Leading into a runtime panic upon failure?)

Reply:
> On a similar note, it is probably worth mentioning type inference. For example:
> 
> ```rust
> fn foo(a: u8, b: u8, c: u8, d: u8, e: u8) { /* ... */ }
> 
> fn bar(buf: &[u8]) {
>     foo(...buf.try_into().unwrap());
> }
> ```
> 
> Is that expected to work because `&[T; 5]` has `TryFrom<&[T]>` and `u8` is `Copy`? What about
> 
> ```rust
> fn bar(buf: &[u8], other: &[u8]) {
>     foo(1, ...buf.try_into().unwrap(), ...other.try_into().unwrap());
> }
> ```
> 
> That probably needs to be forbidden as ambiguous, but that is something that should be spelled out here with some T-Types input.

I'll definitely need to add a subchapter on type inference next to the referred "Type Coercions of Collections" subchapter. 👍🏼

Reply:
> Earlier it is stated that "Tuples, tuple structs, and fixed-size arrays can be unpacked.", but this seems to indicate that anything implementing `Index<Int>` could be unpacked. I think tuples definitely make sense and fixed-length `[T; N]` arrays could make sense, but anything involving `Index` should be out of scope for now because it turns the complexity up quite a bit.
> 
> It is pretty easy to turn slicelike collections into fixed-length arrays anyway, the user can handle that if they need this support.

I've edited this subchapter (now named "Type Coercions and Conversions of Collections") in commit fa28f691bb69de5c2f5a466068687bd5519cb88b to address this. Basically, I think the original writing was erroneous, contradicting with other parts of the text, so thanks for bringing this up. This is also one of the sections (and topics) I'm most unsure about, so please don't hesitate to point out any remaining mistakes. :)